orderbook/src/database/orders.rs

```diff
@@ -168,7 +168,14 @@ const ORDERS_SELECT: &str = "\
     COALESCE(SUM(t.buy_amount), 0) AS sum_buy, \
     COALESCE(SUM(t.sell_amount), 0) AS sum_sell, \
     COALESCE(SUM(t.fee_amount), 0) AS sum_fee, \
-    (COUNT(invalidations.*) > 0 OR o.cancellation_timestamp IS NOT NULL) AS invalidated \
+    (COUNT(invalidations.*) > 0 OR o.cancellation_timestamp IS NOT NULL) AS invalidated, \
+    COALESCE((o.signing_scheme = 'presign' AND ( \
+        SELECT (NOT p.signed) as unsigned \
+        FROM presignature_events p \
+        WHERE o.uid = p.order_uid \
+        ORDER BY p.block_number DESC, p.log_index DESC \
+        LIMIT 1 \
+    )), true) AS presignature_pending \
```

Comment:
Making new discussion so it shows up in the code view again.

The coalesce could be moved to just around the sub query because that is the only part that can be null. Could also do the COUNT thing so you don't need coalesce because with count it won't ever be null but not sure which way of doing it is cleaner.
If the coalesce is not moved only to the inner part that it isn't clear to me what happens in the case where
- signing scheme is != presign
- sub query returns null

I could see Postgres either stopping the evaluation of the AND immediately so presignature_pending is false (correct) or it might evaluate the AND with the second operand being null which turns the whole and into null which is coalesced into true (incorrect).

Reply:
I don't see how to use `COUNT` without filtering `signed = true` rows **after** doing the subquery for the most recent row (otherwise, you will just count the most recent signed presignature row, and not count whether or not the most presignature row is signed - I hope that makes sense).

I could also put `COALESCE` around the subquery only, but I expected that the `o.signing_scheme = 'presign'` would shortcircuit the subquery anyway, and `COALESCE(false, ...)` wouldn't have a performance impact.

Reply:
I don't know if it short circuits that's why I'm suggesting this. But even if it didn't the coalesce around the whole query just isn't needed because `o.signing_scheme = 'presign'` cannot be null so why not coalesce only the second part?

> otherwise, you will just count the most recent signed presignature row
Makes sense.

Reply:
> why not coalesce only the second part?

For sure! When I wrote:

> I could also put COALESCE around the subquery only, but I expected that the

I meant it more like

> Yes, I can put COALESCE around the subquery only. The reason I had done it this way was that I expected ...

Reply:
I quickly checked, and it looks like where the `COALESCE` is doesn't matter for the execution plan.

<details><summary>execution plan</summary>

```
nlordell=# explain SELECT
  COALESCE((o.signing_scheme = 'presign' AND (
    SELECT (NOT p.signed) as unsigned
    FROM presignature_events p
    WHERE o.uid = p.order_uid
    ORDER BY p.block_number DESC, p.log_index DESC
    LIMIT 1
  )), true) as presignature_pending
FROM orders o
GROUP BY o.uid;
                                        QUERY PLAN                                        
------------------------------------------------------------------------------------------
 HashAggregate  (cost=12.88..260.70 rows=230 width=58)
   Group Key: o.uid
   ->  Seq Scan on orders o  (cost=0.00..12.30 rows=230 width=61)
   SubPlan 1
     ->  Limit  (cost=1.06..1.06 rows=1 width=17)
           ->  Sort  (cost=1.06..1.06 rows=1 width=17)
                 Sort Key: p.block_number DESC, p.log_index DESC
                 ->  Seq Scan on presignature_events p  (cost=0.00..1.05 rows=1 width=17)
                       Filter: (o.uid = order_uid)
(9 rows)

nlordell=# explain SELECT
  (o.signing_scheme = 'presign' AND COALESCE((
    SELECT (NOT p.signed) as unsigned
    FROM presignature_events p
    WHERE o.uid = p.order_uid
    ORDER BY p.block_number DESC, p.log_index DESC
    LIMIT 1
  ), true)) as presignature_pending
FROM orders o
GROUP BY o.uid;
                                        QUERY PLAN                                        
------------------------------------------------------------------------------------------
 HashAggregate  (cost=12.88..260.70 rows=230 width=58)
   Group Key: o.uid
   ->  Seq Scan on orders o  (cost=0.00..12.30 rows=230 width=61)
   SubPlan 1
     ->  Limit  (cost=1.06..1.06 rows=1 width=17)
           ->  Sort  (cost=1.06..1.06 rows=1 width=17)
                 Sort Key: p.block_number DESC, p.log_index DESC
                 ->  Seq Scan on presignature_events p  (cost=0.00..1.05 rows=1 width=17)
                       Filter: (o.uid = order_uid)
(9 rows)
```

</details>

With this in mind, I prefer your suggestion of putting `COALESCE` only around the subquery, as it makes it clearer that it is the part of the expression that can be `NULL`.

Reply:
@vkgnosis changed to only `COALESCE` the subquery.